### ISSUE TITLE
[meta] Support project-wide dart format page width configuration

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#56863]
After the [new tall style issue](https://github.com/dart-lang/dart_style/issues/1253), the two issues with the most 👍are both about configuring page width ([1](https://github.com/dart-lang/dart_style/issues/833), [2](https://github.com/dart-lang/dart_style/issues/918)). This has been the #1 feature request for many years with hundreds of users asking for it.

So we are adding support for project-wide page width configuration to the formatter. After weighing a number of ways to author this configuration, the best approach seems to be a combination of two options. Each helps address the problems of the other:

### Look for a surrounding analysis_options.yaml file and let it configure the page width

When formatting a file on disk using `dart format`, we walk up the surrounding directories looking for an `analysis_options.yaml` file. If we find one and it has a section like:

```yaml
formatter:
  page_width: 123
```

Then the file is formatted at that width.

This lets users author the width in one location in a parent directory (likely the root directory of a package), and then all files in that directory tree get formatted the same way. Since `analysis_options.yaml` files also support [an `include` key](https://dart.dev/tools/analysis#the-analysis-options-file) to import another `analysis_options.yaml` file, organizations have a way to have a standardized formatting style shared across all of their projects.

The main problem with using `analysis_options.yaml` (or any other config file) is that code generators and other tools that format code that only exists in memory don't know what config file to use. If the code generator doesn't use the configured width, but the user then runs `dart format` on the whole directory, the generated code will be changed. Likewise, if CI tries to validate that every file is formatted using the configured width, that will fail because the generated stuff is formatted at a different width.

### Allow a comment inside of a source file to indicate the page width for that file

If you put `// dart format width=123` at the top of a file, then the file is formatted at that page width. This is [simple to implement](https://github.com/dart-lang/dart_style/commit/fb00aab5558c807525f2ec134bd67a5f89e31873), works with `dart format`, code generators, and any other tool that wants to format code. It works even without having access to file IO or any knowledge of where the formatted code will end up on disk.

But it's very tedious for users to have to remember to put this in every single file they author. If they forget, they get inconsistently formatted code. No user has asked for this. They want *project-wide* configuration.

### Hybrid approach

We implement *both* of these. If a user has a project-wide page width set in `analysis_options.yaml` and the file has a page width comment, the latter takes precedence (just like a `// @dart=1.2` language version comment takes precedence over the SDK constraint in a pubspec).

Mixing both approaches lets us avoid the problems with each:

*   Tools like `dart format` and IDEs look for a configured project-wide page width in `analysis_options.yaml`. This is implemented in `dart format` [now](https://github.com/dart-lang/dart_style/commit/9c17d3b2ad08a5f86be2b45216b16c71aca24f38) (behind an experiment flag). The analysis_server package, which most Dart IDE integrations use, already has support for working with `analysis_options.yaml` files, so plumbing the page width from that to the files it formats should be straightforward.

    This gives users the project-wide configuration they want.

*   Tools that generate Dart code and then immediately format in memory without having easy access to the surrounding `analysis_options.yaml` file can simply add a `// dart format width=123` comment to the top of the generated code before they format it.

    Since that takes precedence over any project-wide configuration, then formatting that file in memory with no `analysis_options.yaml` file will produce the same results as when the file is later formatted with `dart format` or the IDE. That should mean that things like CI steps that check for formatted files won't end up breaking on generated code.

    This does mean that generated files may have a different width than the user prefers. But these files are fairly rarely read, so that's a tolerable compromise to avoid having every library that works with generated code and the formatter become obligated to support searching for and reading `analysis_options.yaml` (which isn't possible in some contexts like build_runner).

## Implementation tasks

The core formatter work is done now (but behind a flag). The work remaining is largely in other packages:

**TODO: Fill this in.**

**Comment by munificent** [comment]
@dart-lang/analyzer-team, @dart-lang/language-team, @dart-lang/core-package-admins, any other important tools/workflows/packages that should handle this?

**Comment by parlough** [comment]
If you're going to use `formatter` as the top-level key (which makes sense), perhaps that name should be reserved on pub.dev. It seems analyzer plugins (or plugin like tools) often use top-level keys matching their package name for configuration.

**Comment by kevmoo** [comment]
@parlough 's comment reminds me about a similar discussion we had about namespacing "properties" in pubspec.yaml, too.

**Comment by munificent** [comment]
Good idea, @parlough. [Filed an issue](https://github.com/dart-lang/pub-dev/issues/8128) and added it to the list. Thanks!

**Comment by feinstein** [comment]
@munificent could we get a `formatter` flag to configure which files can be excluded from formatting? Maybe a `// dart format exclude` comment also?

This will be useful for skipping generated files entirely in CI runs, so we don't waste compute time formatting generated files.

**Comment by davidmorgan** [comment]
> [@munificent](https://github.com/munificent) could we get a `formatter` flag to configure which files (glob list) can be excluded from formatting? Maybe a `// dart format exclude` comment also?
> 
> This will be useful for skipping generated files entirely in CI runs, so we don't waste compute time formatting generated files.

There is an exclude comment: `// dart format off`, I think this is the right way to do it. Otherwise, passing the list of generated files _everywhere_ is going to be very complicated.

https://github.com/dart-lang/build/issues/3812

**Comment by feinstein** [comment]
> Otherwise, passing the list of generated files everywhere is going to be very complicated.

I mean more something like:

```yaml
# in analysis_options.yaml
  exclude:
    - "**/*.freezed.dart"
    - "**/*.g.dart"
    - "**/*.mock.dart"
```

**Comment by munificent** [comment]
> [@munificent](https://github.com/munificent) could we get a `formatter` flag to configure which files (glob list) can be excluded from formatting? Maybe a `// dart format exclude` comment also?

Potentially, yes. Here's the main issue: https://github.com/dart-lang/dart_style/issues/864

But in general I agree with @davidmorgan that it's better to put these exclusions inside the file itself so that the code generator can handle the exclusion itself without users having to remember to configure it for each of their projects.

**Comment by feinstein** [comment]
I agree with this, but then we need the package maintainer to add this to their generated files, and sometimes they don't and we can't control this. So I think both have their uses.

**Comment by davidmorgan** [comment]
> I agree with this, but then we need the package maintainer to add this to their generated files, and sometimes they don't and we can't control this. So I think both have their uses.

It probably makes sense for `build_runner` to add a standard header; although it won't be a super simple change to land :)

**Comment by dishankjindal1** [comment]
> # in analysis_options.yaml
> formatter:
>   exclude:
>     - "**/*.freezed.dart"
>     - "**/*.g.dart"
>     - "**/*.mock.dart"
> I have this in my CI currently, and it would be nice to be able to remove this:

If this is possible, it will be a game changer as more and more teams are adding code generation to there workflows. Not having to deal with code conflicts will be a major win.

**Comment by Neelansh-ns** [comment]
Looking forward to this 🙌🏻

**Comment by munificent** [comment]
> Looking forward to this 🙌🏻

It has already shipped! :D Dart 3.7 includes the project-wide page width configuration.

This tracking is still open because there is a lingering issue to update the 80 column lint to handle configured page width. But, otherwise, the work is done.
